presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -63,242 +89,265 @@ public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boo
+        return typeSignature.getParameters().stream()
+                .map(TypeSignatureParameter::getKind)
+                .allMatch(kind -> kind == ParameterKind.LONG || kind == ParameterKind.VARIABLE);
+    }
+
+    private Optional<BoundVariables> solve(List<TypeConstraintSolver> constraints)
+    {
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+        for (int i = 0; ; i++) {
+            if (i == SOLVE_ITERATION_LIMIT) {
+                throw new VerifyException(format("SignatureBinder.solve does not converge after %d iterations.", SOLVE_ITERATION_LIMIT));
             }
-            case NAMED_TYPE: {
-                checkState(actual.getKind() == ParameterKind.NAMED_TYPE,
-                        "NAMED_TYPE parameter kind is expected here");
-                NamedTypeSignature expectedNamedTypeSignature = expected.getNamedTypeSignature();
-                NamedTypeSignature actualNamedTypeSignature = actual.getNamedTypeSignature();
-                if (!expectedNamedTypeSignature.getName().equals(actualNamedTypeSignature.getName())) {
-                    return false;
+            boolean changed = false;
+            for (TypeConstraintSolver constraint : constraints) {
+                SolverReturnStatus status = constraint.update(boundVariablesBuilder);
+                if (status == SolverReturnStatus.UNSOLVABLE) {
+                    return Optional.empty();
+                }
+                if (status == SolverReturnStatus.CHANGED) {
+                    changed = true;
                 }
-                TypeSignature expectedTypeSignature = expectedNamedTypeSignature.getTypeSignature();
-                TypeSignature actualTypeSignature = actualNamedTypeSignature.getTypeSignature();
-                Type actualType = typeManager.getType(actualTypeSignature);
-                return bind(expectedTypeSignature, actualType, variableBinder);
             }
-            default:
-                throw new IllegalStateException("Unknown parameter kind: " + expected.getKind());
+            if (changed) {
+                break;
+            }
         }
-    }
 
-    private boolean matchAndBindLongVariable(String variable, Long value, BoundVariables.Builder variableBinder)
-    {
-        if (variableBinder.containsLongVariable(variable)) {
-            Long currentVariableValue = variableBinder.getLongVariable(variable);
-            return value.equals(currentVariableValue);
+        for (TypeConstraintSolver constraint : constraints) {
+            if (!constraint.isSatisfied(boundVariablesBuilder)) {
+                return Optional.empty();
+            }
         }
-        else {
-            variableBinder.setLongVariable(variable, value);
-            return true;
+
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+
+        BoundVariables boundVariables = boundVariablesBuilder.build();
```

Comment:
Plain map does not work very well because of the distinction between type variables and long variables. If having a builder that allows overwriting is surprising to you, maybe we can discuss about changing its name. I would also like to point out that I didn't change the behavior of BoundVariables (now that I undid my change to its getters).